Fix:remove useless time consuming opoeration in MDSUdpEvent

GabrieleManduchi opened the pull request.

Commit:
Fix:remove useless time consuming opoeration in MDSUdpEvent

joshStillerman commented:
Can you provide more explanation of the changes ? Not really understanding the code it seems curious there are as many unprotected static variables in the module.

GabrieleManduchi commented:
When I tested event streaming performance I realized that 3-4 ms are always lost in event communication, even when communicating among thread. After some investigation I realized that this was due the call to getserverbyname() issued by UdpEventGetPort(). So I moved its call to pthread_once already performed for socket creatuion, adding one more static variable sendPort. Anyeay this stati variable, as the other static variables in UdpEvent.c is only accessed in read, and the only write is performed withing pthread_once. Noe commubnitation time dropped from 3-4 ms to 80 us.

GabrieleManduchi commented:
Anyway, wait before merging it. I am now adding the possibility of core assignment for UdpEventAst, that is mandatory if we want to use streams in real-time.

Commit:
Feature:allow assigning core mask to thread listening to events

Extended MDS Event API with  an additional MdsEventAstMask routine. MdsEventAstMask behaves exactly as MdsEventAst but ensures that the created thread is assigned to a core as specified by the core mask. In Event cpp interface a new Event constructor has been added to handle the additional mask argument.

GabrieleManduchi commented:
Now I have added an additional routine in MDS event API: MdsEventAstMask which behaves as MdsEventAst but ensures that the listening thread runs on a core specified in the additional mask argument

Commit:
MOVE include

Commit:
MOVE include

Commit:
MOVE include

Commit:
Check if CPU_SET defined

GabrieleManduchi commented:
retest this please

joshStillerman commented:
Good find and good fix !  This performance is going to be fine for the 
bagel, but will not cut it for our SDN.  We can worry about that later.  
Darren and I are seriously considering doing MARTe time with PTP-v2, and 
using OPC_UA / multicast or OPC_UA/ethernet-TSN for the SDN


On 4/14/21 3:06 AM, GabrieleManduchi wrote:
>
> When I tested event streaming performance I realized that 3-4 ms are 
> always lost in event communication, even when communicating among 
> thread. After some investigation I realized that this was due the call 
> to getserverbyname() issued by UdpEventGetPort(). So I moved its call 
> to pthread_once already performed for socket creatuion, adding one 
> more static variable sendPort. Anyeay this stati variable, as the 
> other static variables in UdpEvent.c is only accessed in read, and the 
> only write is performed withing pthread_once. Noe commubnitation time 
> dropped from 3-4 ms to 80 us.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MDSplus/mdsplus/pull/2282#issuecomment-819283273>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABY5AZIGDFHLXUE7APBHKPLTIU5H5ANCNFSM423IT2DQ>.
>
-- 
*Joshua Stillerman*
Research Engineer
MIT Plasma Science and Fusion Center
617.253.8176
***@***.*** ***@***.***>

Commit:
add attribute unused to mask

GabrieleManduchi commented:
Josh, now I have terminated the fixes and added a new routine in the API  (MDsEventAstMask) that allows defining a CPU mask for the activated thread listening udp messages. This dramatically reduced jitter in my MARTe2 tests.
Could it check and if  you agree approve the PR?
